test/integration/auth/requestheader_test.go

```diff
@@ -0,0 +1,206 @@
+			frontProxyCAFilename := opts.Authentication.RequestHeader.ClientCAFile
+			if err := os.WriteFile(frontProxyCAFilename, frontProxyCA, 0644); err != nil {
+				t.Fatal(err)
+			}
+
+			opts.Authentication.RequestHeader.AllowedNames = append(opts.Authentication.RequestHeader.AllowedNames, "test-aggregated-apiserver")
+			if setUIDHeaders {
+				opts.Authentication.RequestHeader.UIDHeaders = []string{"X-Remote-Uid"}
+			}
+		}
+	}
+
+	for _, tt := range []struct {
+		name   string
+		setUID bool
+	}{
+		{
+			name:   "KAS without UID config",
+			setUID: false,
+		},
+		{
+			name:   "KAS with UID config",
+			setUID: true,
+		},
+	} {
+		t.Run(tt.name, func(t *testing.T) {
+			_, kubeConfig, tearDownFn := framework.StartTestServer(tCtx, t, framework.TestServerSetup{
+				ModifyServerRunOptions: modifyOpts(tt.setUID),
+			})
+			defer tearDownFn()
+
+			// Test an aggregated apiserver client (signed by the new front proxy CA) is authorized
+			extensionApiserverClient, err := kubernetes.NewForConfig(&rest.Config{
+				Host: kubeConfig.Host,
+				TLSClientConfig: rest.TLSClientConfig{
+					CAData:     kubeConfig.TLSClientConfig.CAData,
+					CAFile:     kubeConfig.TLSClientConfig.CAFile,
+					ServerName: kubeConfig.TLSClientConfig.ServerName,
+					KeyData:    frontProxyKey,
+					CertData:   frontProxyClient,
+				},
+			})
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			selfInfo := &authnv1.SelfSubjectReview{}
+			err = extensionApiserverClient.AuthenticationV1().RESTClient().
+				Post().
+				Resource("selfsubjectreviews").
+				VersionedParams(&metav1.CreateOptions{}, scheme.ParameterCodec).
+				Body(&authnv1.SelfSubjectReview{}).
+				SetHeader("X-Remote-Uid", "test-uid").
+				SetHeader("X-Remote-User", "testuser").
+				SetHeader("X-Remote-Groups", "group1", "group2").
+				Do(tCtx).
+				Into(selfInfo)
+			if err != nil {
+				t.Fatalf("failed to retrieve self-info: %v", err)
+			}
+
+			if selfUID := selfInfo.Status.UserInfo.UID; (len(selfUID) != 0) != tt.setUID {
+				t.Errorf("UID should be set: %v, but we got %v", tt.setUID, selfUID)
+			}
+		})
+	}
+
+}
+
+func newTestCAWithClient(caSubject pkix.Name, caSerial *big.Int, clientSubject pkix.Name, subjectSerial *big.Int) (caPEMBytes, clientCertPEMBytes, clientKeyPEMBytes []byte, err error) {
+	ca := &x509.Certificate{
+		SerialNumber:          caSerial,
+		Subject:               caSubject,
+		NotBefore:             time.Now(),
+		NotAfter:              time.Now().Add(24 * time.Hour),
+		IsCA:                  true,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth},
+		KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		BasicConstraintsValid: true,
+	}
+
+	caPrivateKey, err := rsa.GenerateKey(rand.Reader, 4096)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	caBytes, err := x509.CreateCertificate(rand.Reader, ca, ca, &caPrivateKey.PublicKey, caPrivateKey)
```

Comment:
grep for `x509.CreateCertificate` in the codebase - we have multiple helpers in `client-go` and in `test/` for helping with these types of operations.

Reply:
I used some functions for generating the CA and marshalling the pkey. There is nothing for client certs in the lib, unfortunately, so that part has to stay.

Reply:
Why can't we use `k8s.io/kubernetes/test/utils.NewSignedCert`?

Reply:
I must have missed that one as I was focusing mostly on the client-go library once I found the functions there. I'll use that one.

Would be nice to have a library for these things, this pki_utils code appears to be also copied to kubeadm.